### PR TITLE
Adding the CVPR Chesapeake dataset

### PR DESCRIPTION
**Pull request by calebrob6**
Dataset from https://lila.science/datasets/chesapeakelandcover

Some notes:
- `_check_integrity` behaves differently from other places, see #99
- The dataset is made up of *non-overlapping* geotiffs
- The dataset's index is in EPSG:3857 and we index into it with EPSG:3857 boxes, but the actual data is in EPSG:26917 and EPSG:26918.
- In `__getitem__` we grab the single geotiff that the query intersects with, warp the query geom to the CRS of that file, then return the data corresponding to the warped query.

To do:
- [x] upload the updated version of the dataset to lila.science 
- [x] add a mechanism by which users can pick which layers they want, e.g. `["naip-new", "lc"]` should return only those layers
- [x] pyproj dependency

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -27,6 +27,8 @@ pillow>=2.9
 pycocotools
 # pydocstyle 6.1+ required for pyproject.toml support
 pydocstyle[toml]>=6.1
+# pyproj required for fast CRS conversion
```
[the remark this answers is not in the document]
Comment (calebrob6) [reply requirements.txt:30]: 3.1+ is desired actually as it has threadsafe implementation for Transformer and CRS and we likely want to use those in datasets

File: requirements.txt [hunk 2]
```diff
@@ -50,6 +52,8 @@ scikit-learn>=0.18
 segmentation-models-pytorch>=0.2
 # setuptools 30.4+ required for options.packages.find section in setup.cfg
 setuptools>=30.4
+# shapely is required for manipulating and warping geometries
+shapely
```
[the remark this answers is not in the document]
Comment (calebrob6) [reply requirements.txt:56]: I don't have a good min version for shapely, but just tested with 1.7.0 from Jan 2020 and it works.

[the remark this answers is not in the document]
Comment (calebrob6) [reply requirements.txt:56]: 1.6.0 is from 2017 and does _not_ work due to conflicting dependencies with pyproj

File: requirements.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamjstewart) [reply requirements.txt]: Want to use or are currently using? I want to keep the minimum requirements as minimum as possible until we need to increase them.

File: requirements.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamjstewart) [reply requirements.txt]: Hmm, could we change the dependencies too? I usually just pip install the minimum version of everything until it works to see what the actual minimum requirement is.

**Comment by adamjstewart** [comment]
I don't love having to add 2 more deps (seems every PR adds at least 1 dep) but these 2 seem unavoidable in the long run anyway.

**Comment by calebrob6** [comment]
pyproj 3.1 doesn't support python 3.6 @adamjstewart - thoughts on dropping 3.6

**Comment by adamjstewart** [comment]
I would prefer not to drop 3.6 just because of a dependency. Numpy also doesn't support 3.6. Can we just install an older version on 3.6?

**Comment by calebrob6** [comment]
sure how do you do that

**Comment by adamjstewart** [comment]
I would like to test with older versions of pyproj and shapely. Both of these are almost the latest version available. Having the minimum required version so high makes it difficult for people to install.

When I was testing the minimum supported version I basically created a new conda environment with Python 3.6, then pip installed all the deps, then pip installed successively older versions of each package until the tests no longer passed.

**Comment by calebrob6** [comment]
Can you do this if you then? I do not think it is hard for users to install new versions of packages, `pip install ...` covers it.

**Comment by adamjstewart** [comment]
I can once you add unit tests that actually run this code.

No minimum version at all is better than setting the latest version to be the minimum version. Many packages aren't necessarily compatible with the latest version of everything. For example, rasterio does not support the latest version of GDAL. If we decided to require the latest version of GDAL, we wouldn't be able to build rasterio at all.

**Comment by calebrob6** [comment]
Generally true, but that's not the case here. There exists a valid set of our required packages that work with pyproj>=3.0 and shapely>=1.7.0 so I do not see what the problem is.

In any case, I made a test for the shape warping functionality that we're using from pyproj and shapely, I'll go ahead and test some major versions of the two libraries so that this can be merged.

**Comment by adamjstewart** [comment]
> Generally true, but that's not the case here. There exists a valid set of our required packages that work with pyproj>=3.0 and shapely>=1.7.0 so I do not see what the problem is.

If users only installed torchgeo and its dependencies that would be true, but users tend to write their own code and install things like torchgeo into much larger environments where conflicts may be possible.

File: environment.yml [hunk 3]
```diff
@@ -8,10 +8,12 @@ dependencies:
   - numpy
   - pip
   - pycocotools
+  - pyproj>=2.2.0
   - python>=3.6
   - pytorch-gpu>=1.7
   - rarfile>=3
   - rasterio>=1.0.16
+  - shapely>1.5.0
```
Comment (adamjstewart) [inline environment.yml]:
```suggestion
  - shapely>=1.5.0
```

File: tests/test_crs_conversions.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import math
+
+import pyproj
+import shapely.ops
+
+
+def test_crs_with_pyproj() -> None:
```
Comment (adamjstewart) [inline tests/test_crs_conversions.py:10]: This test doesn't actually test anything in `torchgeo`, I would remove it.

File: torchgeo/datasets/__init__.py [hunk 5]
```diff
@@ -57,6 +58,7 @@
     "ChesapeakePA",
     "ChesapeakeVA",
     "ChesapeakeWV",
+    "CVPRChesapeake",
```
Comment (adamjstewart) [inline torchgeo/datasets/__init__.py:61]: Can we call this `ChesapeakeCVPR` to match all of the other dataset names?

File: torchgeo/datasets/cvpr_chesapeake.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (adamjstewart) [inline torchgeo/datasets/cvpr_chesapeake.py:1]: This should be added to `chesapeake.py`

File: torchgeo/datasets/cvpr_chesapeake.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+
+    * https://doi.org/10.1109/cvpr.2019.01301
+    """
+
+    url = "https://lilablobssc.blob.core.windows.net/lcmcvpr2019/cvpr_chesapeake_landcover.zip"  # noqa: E501
+    filename = "cvpr_chesapeake_landcover.zip"
+    md5 = "0ea5e7cb861be3fb8a06fedaaaf91af9"
+
+    crs = CRS.from_epsg(3857)
+    res = 1
```
Comment (adamjstewart) [inline torchgeo/datasets/cvpr_chesapeake.py:43]: What are these actually used for?

File: torchgeo/datasets/cvpr_chesapeake.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+        cache: bool = True,
+        download: bool = False,
+        checksum: bool = False,
+    ) -> None:
+        """Initialize a new Dataset instance.
+
+        Args:
+            root: root directory where dataset can be found
+            split: a string in the format "{state}-{train,val,test}" indicating the
+                subset of data to use
```
Comment (adamjstewart) [inline torchgeo/datasets/cvpr_chesapeake.py:86]:
```suggestion
                subset of data to use, for example "ny-train"
```
I assume this is what is meant?

File: environment.yml [hunk 9]
```diff
@@ -8,10 +8,12 @@ dependencies:
   - numpy
   - pip
   - pycocotools
+  - pyproj>=2.2.0
   - python>=3.6
   - pytorch-gpu>=1.7
   - rarfile>=3
   - rasterio>=1.0.16
+  - shapely>1.3.0
```
Comment (adamjstewart) [inline environment.yml:16]:
```suggestion
  - shapely>=1.3.0
```

**Comment by adamjstewart** [comment]
Might also need to add deps to `docs/requirements.txt` (added in `main`). I'm hoping we don't have to keep this file forever. This can either be done in this PR if you rebase or after its merged and the docs start failing to build.

File: torchgeo/datasets/cvpr_chesapeake.py (hunk 7 above)
Comment (calebrob6) [reply torchgeo/datasets/cvpr_chesapeake.py:43]: defining the properties from GeoDataset

File: torchgeo/datasets/cvpr_chesapeake.py (hunk 8 above)
Comment (calebrob6) [reply torchgeo/datasets/cvpr_chesapeake.py:86]: yes

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
Just realized we forgot to add this dataset to the docs: https://github.com/microsoft/torchgeo/commit/8ab42118287ca15040dc74d101b830588992e258

**Pull request merged**